### DISCUSSION TITLE
What is the difference between addStack, insertStack and setStack methods in the PlayerInventory class, and when are they called?

### DISCUSSION THREAD

**Discussion by TechPro424** [FabricMC/fabric-api#2834] [Mod Dev Support] [answered]
What is the difference between `addStack`, `insertStack` and `setStack` methods in the `PlayerInventory` class, and when is each method called?

**Comment by TechPro424** [comment]
bump

**Comment by KosmX** [comment, marked as answer]
`addStack` is an internal function, you shouldn't use that.  
If you want to *give* the player some items, use `insertStack`. That will pick up the stack if there is space for the stack, like what `/give` does.  
`setStack` replaces the selected stack with a new one. maybe replace a used tool with a broken one.
